src/com/twitter/intellij/pants/execution/PantsMakeBeforeRun.java

```diff
@@ -313,6 +313,16 @@ public void onTextAvailable(ProcessEvent event, Key outputType) {
   public PantsExecuteTaskResult executeCompileTask(@NotNull Project currentProject,
                                                    @NotNull Set<String> targetAddressesToCompile,
                                                    boolean useCleanAll) {
+
+    ApplicationManager.getApplication().invokeAndWait(new Runnable() {
+      @Override
+      public void run() {
+        /* Force cached changes to disk so {@link com.twitter.intellij.pants.file.FileChangeTracker} can mark the project dirty. */
+        FileDocumentManager.getInstance().saveAllDocuments();
+        currentProject.save();
```

Comment:
This is the crux of the fix, because the bug was that files were saved after the compile action was called.

Reply:
Does invokeAndWait here have any potential for causing a UI hang, or is this all on a background thread?

Reply:
I don't think the code runs on UI thread (inside and outside of `run()`), so no.